### PR TITLE
sysroot: Stabilize deployment finalization, add API

### PR DESCRIPTION
**Pull request by cgwalters**
It's about time we do this; deployment finalization locking is a useful feature.  An absolutely key thing here is that we've slowly been moving towards the deployments as the primary "source of truth".

Specifically in bootc for example, we will GC container images not referenced by a deployment.

This is then neecessary to support a "pull but don't apply automatically" model.

Closes: https://github.com/ostreedev/ostree/issues/3025

### CODE REVIEW

**Review by ericcurtin — APPROVED** [review]
LGTM

File: src/libostree/ostree-sysroot-deploy.c [hunk 1]
```diff
@@ -3722,11 +3726,22 @@ _ostree_sysroot_finalize_staged_inner (OstreeSysroot *self, GCancellable *cancel
...
-    return FALSE;
-  if (errno == 0)
+  gboolean locked = false;
+  (void)g_variant_lookup (self->staged_deployment_data, _OSTREE_SYSROOT_STAGED_KEY_LOCKED, "b",
+                          &locked);
+  if (locked)
+    g_debug ("staged is locked via metadata");
+  else
+    {
+      if (!glnx_fstatat_allow_noent (AT_FDCWD, _OSTREE_SYSROOT_RUNSTATE_STAGED_LOCKED, NULL, 0,
```
Comment (jlebon) [inline src/libostree/ostree-sysroot-deploy.c:3736]: As part of stabilization, shouldn't we make this file path public? And document how to unlock it.

Also, should we add `ostree admin finalize-staged --unlock`?

Comment (cgwalters) [reply src/libostree/ostree-sysroot-deploy.c:3784]: Right thanks, this was not fully baked.  I've updated this more, though needs some testing.

Comment (cgwalters) [reply src/libostree/ostree-sysroot-deploy.c:3784]: > As part of stabilization, shouldn't we make this file path public? 

Here we're moving the locked state into the staged deployment data itself (which would have avoided the earlier problem) but it does mean we need an API and CLI to mutate that state.

File: man/ostree-admin-unlock-finalization.xml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    </refsynopsisdiv>
+
+    <refsect1>
+        <title>Description</title>
+
+        <para>
+            This command requires a staged deployment.  If it has been "finalization locked",
+            which means it will not be queued for the next boot by default, then this command
+            will unlock it.  If it is already finalization-unlocked, then this command will
+            be a no-op.
```
Comment (jlebon) [inline man/ostree-admin-unlock-finalization.xml:63]: The API implemented below isn't idempotent as described here.

File: man/ostree-admin-unlock-finalization.xml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public
+License along with this library. If not, see <https://www.gnu.org/licenses/>.
+-->
+
+<refentry id="ostree">
+
+    <refentryinfo>
+        <title>ostree admin unlock-finalization</title>
```
Comment (jlebon) [inline man/ostree-admin-unlock-finalization.xml:27]: Should we doc the `--lock-finalization` option in the `ostree admin deploy` page and  cross-reference this manpage/command?

**Review by jlebon — COMMENTED** [review]
Looks sane overall! CI failure looks legit.

**Comment by cgwalters** [comment]
> CI failure looks legit.

Hmm yes possibly, though I'm not reproducing this here.  One thing this really brings to the fore is the huge messiness around gpg verification vs direct deployments and how `ostree admin status` bails out with an command-fatal error versus what rpm-ostree does in just showing the error as part of state gathering.

File: man/ostree-admin-deploy.xml [hunk 4]
```diff
@@ -121,6 +121,16 @@ License along with this library. If not, see <https://www.gnu.org/licenses/>.
                 </para></listitem>
             </varlistentry>
 
+            <varlistentry>
+                <term><option>--lock-finalization</option></term>
+
+                <listitem><para>
+                    The deployment will not be "finalized" by default on shutdown; to later
+                    queue it, use <literal>ostree admin unlock-finalization</literal>.
```
Comment (jlebon) [inline man/ostree-admin-deploy.xml:129]:
```suggestion
                    queue it, use <literal>ostree admin lock-finalization --unlock</literal>.
```

File: man/ostree-admin-lock-finalization.xml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    <refsect1>
+        <title>Description</title>
+
+        <para>
+            This command requires a staged deployment.  By default, this command ensures the deployment
+            will be set into a "finalization locked" state, which means it will not be queued for the next boot by default.
+        </para>
+        <para>
+            This is the same as the <literal>--lock-finalization</literal> argument for <literal>ostree admin deploy</literal>.
```
Comment (jlebon) [inline man/ostree-admin-lock-finalization.xml:64]: Feels like we should be more explicit about the race condition of not doing this. Maybe e.g.

```suggestion
            This is the same as the <literal>--lock-finalization</literal> argument for <literal>ostree admin deploy</literal>, which is the recommended way to use this feature in a race-free way.
```

?

**Review by jlebon — APPROVED** [review]
Nice!

Some docs comments, but fine to leave as follow-ups to not waste CI.

File: man/ostree-admin-deploy.xml (hunk 4 above)
Comment (cgwalters) [reply man/ostree-admin-deploy.xml:129]: Good catch, thanks for the careful review!

**Comment by cgwalters** [comment]
Followup in https://github.com/ostreedev/ostree/pull/3100

**Pull request merged**